### PR TITLE
Issue #243 contrafactual accounts

### LINKED ISSUE #243: Contrafactual accounts
Regarding [this PR in the JS impl](https://github.com/AdExNetwork/adex-validator/pull/250) we need to make the appropriate changes by removing the Identity contract calls and use the relay instead.

### PR DESCRIPTION
**Pull request by elpiel**
Resolves #243 

Changes applied from: https://github.com/AdExNetwork/adex-validator/pull/250

### CODE REVIEW

File: adapter/src/ethereum.rs [hunk 1]
```diff
@@ -276,6 +266,49 @@ impl Adapter for EthereumAdapter {
...
+            .get(&relay_url)
+            .send()
+            .and_then(|res: Response| res.json())
+            .await
+            .map_err(|_| map_error("Fetching privileges failed"))?;
+
+        let privilege = identities_owned
+            .get(identity)
+            .copied()
+            .unwrap_or_else(|| 0_u8);
```
Comment (samparsky) [inline adapter/src/ethereum.rs]: It shouldn't return `0_u8` if it fails to unwrap it, it means the identity isn't allowed. it should return an error `insufficient privilege`

File: adapter/src/ethereum.rs [hunk 2]
```diff
@@ -276,6 +266,49 @@ impl Adapter for EthereumAdapter {
...
+    pub fn new(relayer_url: &str) -> Result<Self, reqwest::Error> {
+        let client = Client::builder().build()?;
+
+        Ok(Self {
+            relayer_url: relayer_url.to_string(),
+            client,
+        })
+    }
+
+    pub async fn get_privilege(&self, identity: &[u8; 20]) -> Result<u8, AdapterError> {
```
Comment (samparsky) [inline adapter/src/ethereum.rs]: we should use `ValidatorId` for identity

File: adapter/src/ethereum.rs [hunk 3]
```diff
@@ -276,6 +266,49 @@ impl Adapter for EthereumAdapter {
...
+        use reqwest::Response;
+        use std::collections::HashMap;
+
+        let relay_url = format!(
+            "{}/identity/by-owner/{}",
+            self.relayer_url,
+            hex::encode(identity)
```
Comment (samparsky) [inline adapter/src/ethereum.rs]: identity is `0x` prefixed

File: adapter/src/ethereum.rs (hunk 1 above)
Comment (elpiel) [reply adapter/src/ethereum.rs]: `0` is no privileges, and if you've seen the code that uses this function, you will see that we check if it's `> 0`

File: adapter/src/ethereum.rs (hunk 2 above)
Comment (elpiel) [reply adapter/src/ethereum.rs]: Sounds good!

File: adapter/src/ethereum.rs (hunk 3 above)
Comment (elpiel) [reply adapter/src/ethereum.rs]: Yeah I am still testing it, couldn't do it in Thursday.

File: adapter/src/ethereum.rs (hunk 1 above)
Comment (elpiel) [reply adapter/src/ethereum.rs]: We could make an enum, but not sure if you'll like that

Comment (samparsky) [reply adapter/src/ethereum.rs]: We don't need the privilege levels anymore, we can just return the error if we don't find it

File: adapter/src/ethereum.rs (hunk 3 above)
Comment (elpiel) [reply adapter/src/ethereum.rs]: This is fixed! Also I had to pass the `VerifiedPayload.from` and not the identity here.

**Pull request merged**